connector/jobrunner/channels.py

```diff
@@ -0,0 +1,602 @@
+    >>> j4 < j5 < j3
+    True
+
+    j6 has same date_created and priority as j5 but a smaller eta
+    >>> j6 = ChannelJob(None, 6, seq=0, date_created=5, priority=9, eta=2)
+    >>> j6 < j4 < j5
+    True
+
+    Here is the complete suite:
+    >>> j6 < j4 < j5 < j3 < j1 < j2
+    True
+
+    j0 has the same properties as j1 but they are not considered
+    equal as they are different instances
+    >>> j0 = ChannelJob(None, 1, seq=0, date_created=1, priority=9, eta=None)
+    >>> j0 == j1
+    False
+    >>> j0 == j0
+    True
+    """
+
+    def __init__(self, db_name, channel, uuid,
+                 seq, date_created, priority, eta):
+        self.db_name = db_name
+        self.channel = channel
+        self.uuid = uuid
+        self.seq = seq
+        self.date_created = date_created
+        self.priority = priority
+        self.eta = eta
+
+    def __repr__(self):
+        return "<ChannelJob %s>" % self.uuid
+
+    def __eq__(self, other):
+        return id(self) == id(other)
+
+    def __hash__(self):
+        return id(self)
+
+    def __cmp__(self, other):
+        if self.eta and not other.eta:
+            return -1
+        elif not self.eta and other.eta:
+            return 1
+        else:
+            return (cmp(self.eta, other.eta) or
+                    cmp(self.priority, other.priority) or
+                    cmp(self.date_created, other.date_created) or
+                    cmp(self.seq, other.seq))
+
+
+class ChannelQueue:
+    """A channel queue is a priority queue for jobs that returns
+    jobs with a past ETA first.
+
+    >>> q = ChannelQueue()
+    >>> j1 = ChannelJob(None, 1, seq=0, date_created=1, priority=1, eta=10)
+    >>> j2 = ChannelJob(None, 2, seq=0, date_created=2, priority=1, eta=None)
+    >>> j3 = ChannelJob(None, 3, seq=0, date_created=3, priority=1, eta=None)
+    >>> q.add(j1)
+    >>> q.add(j2)
+    >>> q.add(j3)
+    >>> q.pop(now=1)
+    <ChannelJob 2>
+    >>> q.pop(now=11)
+    <ChannelJob 1>
+    >>> q.pop(now=12)
+    <ChannelJob 3>
+    """
+
+    def __init__(self):
+        self._queue = PriorityQueue()
+        self._eta_queue = PriorityQueue()
+
+    def __len__(self):
+        return len(self._eta_queue) + len(self._queue)
+
+    def __contains__(self, o):
+        return o in self._eta_queue or o in self._queue
+
+    def add(self, job):
+        if job.eta:
+            self._eta_queue.add(job)
+        else:
+            self._queue.add(job)
+
+    def remove(self, job):
+        self._eta_queue.remove(job)
+        self._queue.remove(job)
+
+    def pop(self, now):
+        if len(self._eta_queue) and self._eta_queue[0].eta <= now:
+            return self._eta_queue.pop()
+        else:
+            return self._queue.pop()
+
+
+class Channel:
+    """A channel for jobs, with a maximum number of workers.
+
+    Job channels are joined in a hierarchy down to the root channel.
+    When a job channel has free workers, jobs are dequeued, marked
```

Comment:
I find the term `workers` confusing here as this word is used as for a different meaning in Odoo (we could argue that the `workers` are how many odoo workers we can speak to at the same time but it would be a bit far-fetched imo)
Would `capacity` be better? Other ideas? And if no better idea I can live with `workers` ;-)